### ISSUE TITLE
Replace vs Additive scene

### ISSUE THREAD

**Issue by notoriousb1t** [steel-scene/steel-scene#6]
In the current iteration, it uses a replace behavior, @shshaw what do you think of using additive behavior instead of replace, or supporting both in the future?

In additive, the states would be more like stacking effects instead of the destination values.  I'm not sure which is more intuitive

**Comment by shshaw** [comment]
You mean with additive, if you're at `x: 100` and you do a state change to `x: 50`, you actually end up at `x: 150`?

**Comment by notoriousb1t** [comment]
yes.  instead of having an initial state, you would have stacking modifiers.  Kind of like having buffs/dots in an MMO

**Comment by shshaw** [comment]
I think that would be a bit confusing as default behavior, but a GSAP-like syntax of `x: '+=100'` could allow clear support & distinction between both additive and replace.

**Comment by notoriousb1t** [comment]
I hadn't considered that.  I have code from just-animate that already can parse += units, so I can pull that in.

How do you think we should handle delays and staggering?

**Comment by shshaw** [comment]
That'd be great!

Regarding staggering & delay, here's my first take:

For staggering, allowing `ref` for `s-target` to receive multiple nodes/targets (assuming that wasn't planned already) and a `stagger` attribute that would determine delays between the targets.

```
    <s-state name="hiddenLeft" duration="1000">
        <s-target ref=".boxes" stagger="500" opacity="0" x="-50px" /><!-- class selector receives multiple elements -->
        <s-target ref="#box1, #box2" stagger="500" opacity="0" x="-50px" /><!-- comma separated selectors -->
    </s-state>
```

My thought with the syntax has been the parent properties/attributes are inherited by the children unless overwritten; akin to default options or prototypes. There may need to be certain exclusions or white-listing, like only the transition/timing style properties `duration`, `delay`, `easing`, `transition` (reference to an `s-transition` ID), etc.

That said, a property like `stagger` would be confusing if you wanted to apply it to the targets within a state, so `staggerTargets` could allow for effects like this:

```
    <s-state name="hiddenLeft" duration="1000" staggerTargets="500">
        <s-target ref="#box1" opacity="0" x="-50px" /><!-- triggers at 0ms -->
        <s-target ref="#box2" opacity="0" x="-50px" /><!-- triggers at 500ms -->
    </s-state>
```

There may also be need for `s-group` to allow for easily trickling down properties to separate targets within a state.

```
    <s-state name="hiddenLeft" duration="1000" staggerTargets="500">
        <s-target ref="#box1" opacity="0" x="-50px" /><!-- triggers at 0ms -->
        <s-target ref="#box2" opacity="0" x="-50px" /><!-- triggers at 500ms -->
    </s-state><!-- state transition complete at 1500ms -->
```


For delay, the same ideas apply:

```
    <s-state name="hiddenLeft" duration="1000">
      <s-group staggerTargets="500">
        <s-target ref="#box1" opacity="0" x="-50px" /><!-- triggers at 0ms -->
        <s-target ref="#box2" opacity="0" x="-50px" /><!-- triggers at 500ms -->
      </s-group>
      <s-target ref="#box3" scaleX="3" /><!-- triggers at 0, complete at 1000ms -->
    </s-state><!-- state transition complete at 1500ms -->
```

However assuming parent properties are inherited as described above, what would be the expected behavior in an instance like this?

```
    <s-state name="hiddenLeft" duration="1000" delay="500">
        <s-target ref="#box1" opacity="0" x="-50px" /><!-- triggers at 500 -->
        <s-target ref="#box2" opacity="0" x="-50px" delay="250" /><!-- Should this trigger at 250 or 750? -->
    </s-state>
```

Combining delay + stagger:

```
    <s-state name="hiddenLeft" duration="1000" >
        <s-target ref="#box1, #box2, #box3" opacity="0" x="-50px" delay="500" stagger="100" /><!-- #box1 triggers at 500, #box2 triggers at 600, #box3 triggers at 700. -->
    </s-state><!-- state transition complete after 1700ms -->
```

Does that all make sense?

**Comment by notoriousb1t** [comment]
These are great ideas.  I am a little wary of staggerTargets because it implies that element order is significant.  Maybe when we get to keyframes it will be significant anyway?

I didn't plan to handle multiple targets per ref (because of experience with this in just-animate), but we probably have to in order to support repeaters.

I like the idea of allowing most properties to cascade.  That should make it pretty intuitive and is a very simple approach on our part.

**Comment by shshaw** [comment]
> support repeaters
What do you mean here?

I think ref almost has to support multiple targets due to the nature of selectors.

In addition to target order,  we'd need to figure out how to handle `staggerTargets` for `s-target`s that reference multiple nodes (classes, comma separated values, etc). Would each element in the `s-target` count in terms of staggering, or would that `s-target`'s elements all receive the same stagger delay?

**Comment by notoriousb1t** [comment]
>support repeaters
What do you mean here?

I mean lists like v-for.  

>In addition to target order, we'd need to figure out how to handle staggerTargets for s-targets that reference multiple nodes (classes, comma separated values, etc). Would each element in the s-target count in terms of staggering, or would that s-target's elements all receive the same stagger delay?

I think it would be more predictable to have staggerTargets treat each s-target as a group and apply the same stagger delay to multiple elements.
